src/replication/ng/flow/flow.go

```diff
@@ -0,0 +1,229 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package flow
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/replication/ng/model"
+
+	"github.com/goharbor/harbor/src/replication/ng/adapter"
+	"github.com/goharbor/harbor/src/replication/ng/execution"
+	"github.com/goharbor/harbor/src/replication/ng/registry"
+	"github.com/goharbor/harbor/src/replication/ng/scheduler"
+)
+
+// Controller controls the replication flow
+type Controller interface {
+	// Start a replication according to the policy and returns the
+	// execution ID and error
+	StartReplication(policy *model.Policy) (int64, error)
+	// Stop the replication specified by the execution ID
+	StopReplication(int64) error
+}
+
+// NewController returns an instance of a Controller
+func NewController(registryMgr registry.Manager,
+	executionMgr execution.Manager, scheduler scheduler.Scheduler) Controller {
+	return &defaultController{
+		registryMgr:  registryMgr,
+		executionMgr: executionMgr,
+		scheduler:    scheduler,
+	}
+}
+
+// defaultController is the default implement for the Controller
+type defaultController struct {
+	registryMgr  registry.Manager
+	executionMgr execution.Manager
+	scheduler    scheduler.Scheduler
+}
+
+// Replicate according the to policy ID
+func (d *defaultController) StartReplication(policy *model.Policy) (int64, error) {
```

Comment:
Suggest to split the method into multi meaningful stages and then reduce the combined methods like `initialize` and `createNamespace` etc.. The body of this method is too large.

Reply:
Can we create a flow object to cover those stages?